docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,348 @@
+| `RESTRICTED_TABLES` | A special privilege which means that the SEM hidden table semantic doesn’t apply. | It is intended for the CloudAdmin user in DbaaS. | 
+
+### Parser Changes
+
+The parser already supports `DYNAMIC` privileges. This can be confirmed by the following patch to TiDB, where they are sent as the static type of `ExtendedPriv`:
+
+```
+diff --git a/planner/core/planbuilder.go b/planner/core/planbuilder.go
+index 90d5b9e82..d1644ee83 100644
+--- a/planner/core/planbuilder.go
++++ b/planner/core/planbuilder.go
+@@ -2298,6 +2298,11 @@ func collectVisitInfoFromGrantStmt(sctx sessionctx.Context, vi []visitInfo, stmt
+ 
+        var allPrivs []mysql.PrivilegeType
+        for _, item := range stmt.Privs {
++
++               if item.Priv == mysql.ExtendedPriv {
++                       fmt.Printf("### Attempting to set DYNAMIC privilege: %s\n", item.Name)
++               }
++
+                if item.Priv == mysql.AllPriv {
+                        switch stmt.Level.Level {
+                        case ast.GrantLevelGlobal:
+```
+
+This results in the following written to the log file:
+```
+mysql> grant acdc on *.* to u1;
+ERROR 8121 (HY000): privilege check fail
+
+### Attempting to set DYNAMIC privilege: acdc
+```
+
+It might be possible that changes are still required if `ExtendedPriv` is not supported in all contexts (REVOKE, etc).
+
+Note that creating a role with the same name as a DYNAMIC privilege is supported. A `GRANT` statement can be attributed to ROLES when it omits the ON *.* syntax. Thus:
+
+```
+GRANT BINLOG_ADMIN TO u1; // grants the role binlog_admin
+GRANT BINLOG_ADMIN ON *.* TO u1; // grants the dynamic privilege binlog_admin
+```
+
+This same nuance applies to MySQL.
+
+### Documentation Plan
+
+The statement reference pages for each of the affected metadata commands will need to be updated to describe dynamic privileges.
+
+There will also need to be documentation specific to `DYNAMIC` privileges to describe how it works, and the purpose of fine-grained access control.
+
+## Test Design
+
+Testing dynamic privileges is a little bit complex because of the various ways privileges can be inherited by users:
+
+* Direct `GRANT` to the user
+* Granting to a role that the user inherits.
+
+### Functional Tests
+
+Unit tests will be added to cover the semantics around role/dynamic privilege precedence, including logical restoring in a different order.
+
+Unit tests will also cover each of the "initial set of Dynamic privileges". Tests will include both directly assigning the privileges and assigning via a ROLE (RBAC).
+
+Integration testing needs to test with global kill enabled/disabled.
+
+### Scenario Tests
+
+The use-cases required by the DBaaS team should be validated when combined with `security-enhanced-mode`. They are:
+
+| Account Name | root | cloudAdmin |
+| --------------- | --------------- | --------------- |
+| Backup & Restore to cloud | Y | Y |
+| File privilege | N | N |
```

Comment:
Maybe this scenario is missed.
I think we should limit the SUPER’ ability to read/write file, like select into file or backup into file.

Reply:
`SUPER` can't read and write a file though? It requires `FILE` privilege, which I verified is correctly checked in TiDB.

But I don't recommend we grant `FILE` to DBaaS users. See https://github.com/pingcap/tidb-test/issues/1154

Reply:
`SUPER` can write a file on MySQL 8.0 with `--secure-file-priv` enabled.

I do not know TiDB's behavior, but there must be a problem when SUPER can write a file.

Reply:
Here is a testcase in 8.0.23 which shows that `SUPER` can not write to `/tmp` (secure-file-priv is set to /tmp):

```sql
mysql [localhost:8023] {root} ((none)) > SELECT 'aa' INTO OUTFILE '/tmp/aa.txt';
Query OK, 1 row affected (0.00 sec)

mysql [localhost:8023] {root} ((none)) > SHOW GRANTS;
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Grants for root@localhost                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| GRANT SELECT, INSERT, UPDATE, DELETE, CREATE, DROP, RELOAD, SHUTDOWN, PROCESS, FILE, REFERENCES, INDEX, ALTER, SHOW DATABASES, SUPER, CREATE TEMPORARY TABLES, LOCK TABLES, EXECUTE, REPLICATION SLAVE, REPLICATION CLIENT, CREATE VIEW, SHOW VIEW, CREATE ROUTINE, ALTER ROUTINE, CREATE USER, EVENT, TRIGGER, CREATE TABLESPACE, CREATE ROLE, DROP ROLE ON *.* TO `root`@`localhost` WITH GRANT OPTION                                                                                                                                                                                                                    |
| GRANT APPLICATION_PASSWORD_ADMIN,AUDIT_ADMIN,BACKUP_ADMIN,BINLOG_ADMIN,BINLOG_ENCRYPTION_ADMIN,CLONE_ADMIN,CONNECTION_ADMIN,ENCRYPTION_KEY_ADMIN,FLUSH_OPTIMIZER_COSTS,FLUSH_STATUS,FLUSH_TABLES,FLUSH_USER_RESOURCES,GROUP_REPLICATION_ADMIN,INNODB_REDO_LOG_ARCHIVE,INNODB_REDO_LOG_ENABLE,PERSIST_RO_VARIABLES_ADMIN,REPLICATION_APPLIER,REPLICATION_SLAVE_ADMIN,RESOURCE_GROUP_ADMIN,RESOURCE_GROUP_USER,ROLE_ADMIN,SERVICE_CONNECTION_ADMIN,SESSION_VARIABLES_ADMIN,SET_USER_ID,SHOW_ROUTINE,SYSTEM_USER,SYSTEM_VARIABLES_ADMIN,TABLE_ENCRYPTION_ADMIN,XA_RECOVER_ADMIN ON *.* TO `root`@`localhost` WITH GRANT OPTION |
| GRANT PROXY ON ''@'' TO 'root'@'localhost' WITH GRANT OPTION                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
3 rows in set (0.00 sec)

mysql [localhost:8023] {root} ((none)) > CREATE USER sonly;
Query OK, 0 rows affected (0.01 sec)

mysql [localhost:8023] {root} ((none)) > GRANT SUPER ON *.* TO sonly;
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql [localhost:8023] {root} ((none)) > SHOW WARNINGS;
+---------+------+----------------------------------------------+
| Level   | Code | Message                                      |
+---------+------+----------------------------------------------+
| Warning | 1287 | The SUPER privilege identifier is deprecated |
+---------+------+----------------------------------------------+
1 row in set (0.00 sec)

mysql [localhost:8023] {root} ((none)) > ^DBye
morgo@ubuntu:~/sandboxes/msb_8_0_23$ ./use -usonly -p
Enter password: 
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 10
Server version: 8.0.23 MySQL Community Server - GPL

Copyright (c) 2000, 2021, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql [localhost:8023] {sonly} ((none)) > SELECT 'bb' INTO OUTFILE '/tmp/bb.txt';
ERROR 1227 (42000): Access denied; you need (at least one of) the FILE privilege(s) for this operation
```